README.md
# iwf-api


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### next

- A working scraper



iwf/event.py
from .core import *
import json


class Event(object):
    def __init__(self, keywords=[], *args):
        self.keywords = keywords

    def _craft_bodyweight_url(new_or_old=None):
        if new_or_old == "old":
            return BASE_URL + OLD_BW_EVENTS_URL
        else:
            return BASE_URL + EVENTS_URL

    def _craft_url(
        self,
        url=None,
        new_or_old=None,
        year=None,
        nation=None,
        event_type=None,
        age_group=None,
    ):

        if url and is_event(url):
            r = requests.get(url, headers=HEADERS)

        else:
            search_url = self._craft_bodyweight_url(new_or_old)
            filters = []
            if year:
                search_url += YEAR_URL
            else:
                if event_type:
                    filters.append(EVENT_TYPE_URL + event_type)
                elif age_group:
                    filters.append(EVENT_AGE_URL + age_group)
                elif nation:
                    filters.append(EVENT_NATION_URL + nation)
            # search_url += "&".join(filters)
            for i in range(len(filters)):
                if i == 0:
                    search_url += "?" + filters[0]
                search_url += "&" + filters[i]
        return

    def _load_event_page(self, search_url):
        r = requests.get(search_url, headers=HEADERS)

        html = r.text
        return BeautifulSoup(html, "lxml")

    _load_event_page(year=2022)

    def _scrape_event_type(self):
        return

    def _scrape_event_age_group(self):
        return

    def _scrape_event_nation(self):
        return

    def _scrape_event_div_card(self):
        """
        <div class="cards">
            <a href="?event_id=529" class="card">
            ...
        </div>
        """
        return

    def _scrape_event_info(self):  # (self, li or soup_data)
        data = {
            "name": None,  # string
            "url": None,  # string
            "location": None,  # string
            "date": None,  # string
        }
        return

    def get_events(
        self,
        year=None,
        new_or_old=None,
        bodyweight=None,
        quantity=None,
        infinity=False,
        sort_by=None,
        nation=None,
        event_type=None,
        age_group=None,
    ):
        return 0


iwf/helper.py
# from iwf.core import *
# from bs4 import BeautifulSoup
# from urllib.parse import urlencode
# from threading import Thread
# from queue import Queue

# <---------------------------------Validations------------------------------->


# def isResult(url):
#     """
#     Validates event url.
#     Example: https://iwf.sport/results/results-by-events/?event_id=529
#     """
#     return True if RESULT_URL in url else False
# 
# 
# def isAthleteBio(url):
#     """
#     Validate athlete url.
#     Example: https://iwf.sport/weightlifting_/athletes-bios/?athlete=ilyin-ilya-1988-05-24&id=7895
#     """
#     return True if ATHLETE_BIO_URL in url else False
# 
# 
# def _loadNewBodyweightEventsPage():
#     """
#     Loads the page for new bodyweight category
#     """
#     r = requests.get(BASE_URL + EVENTS_URL, headers=HEADERS)
#     html = r.text
#     return BeautifulSoup(html, "lxml")
# 
# 
# def _loadOldBodyweightEventsPage():
#     """
#     Loads the page for new bodyweight category
#     """
#     r = requests.get(BASE_URL + OLD_BW_EVENTS_URL, headers=HEADERS)
#     html = r.text
#     return BeautifulSoup(html, "lxml")
# 
# 
# def _scrapeSelectYears(page):
#     """
#     Scrapes data for new or old bodyweight page
#     """
#     select_option = page.findAll("select", {"name": "event_year"})[0]
#     options = select_option.findAll("option")
#     years = []
#     for item in options:
#         years.append(item.get_text())
# 
#     return years
# 
# 
# def getYears():
#     """
#     Gets all years available.
#     New bodyweight years not needed since old bodyweight includes them
#     """
# 
#     # old_bw = []
#     # for year in _scrapeSelectYears(_loadOldBodyweightEventsPage):
#     #     if year <= 2018:
#     #         old_bw.append(year)
# 
#     # new_events_years = [_scrapeSelectYears(_loadNewBodyweightEventsPage())]
#     old_events_years = [_scrapeSelectYears(_loadOldBodyweightEventsPage())]
# 
# 
#     for i in range(len(old_events_years)):
#         if old_events_years[i]:
#             print(old_events_years[i])
# 
#     return 0
# 
# 
# """
# print(_scrapeSelectYears(_loadNewBodyweightEventsPage()))
# print(_scrapeSelectYears(_loadOldBodyweightEventsPage()))
# """
# 
# print(getYears())


iwf/result.py
class Result(object):
    def _scrape_result_info(self):  # (self, li or soup_data)
        data = {
            "name": None,  # string
            "birthdate": None,  # string
            "nation": None,  # string
            "athlete_url": None,  # string
            "category": None,  # string
            "bodyweight": None,  # string
            "group": None,  # string
            "snatch1": None,  # string
            "snatch2": None,  # string
            "snatch3": None,  # string
            "snatch": None,  # string
            "jerk1": None,  # string
            "jerk2": None,  # string
            "jerk3": None,  # string
            "jerk": None,  # string
            "total": None,  # string
            "rank_sn": None,  # string
            "rank_cj": None,  # string
            "rank": None,  # string
        }

    def get_results(self):
        return


iwf/core.py
# -*- coding: utf-8 -*-

import requests
import time

from bs4 import BeautifulSoup
from urllib.parse import urlencode
from threading import Thread
from queue import Queue


BASE_URL = "https://iwf.sport"
HEADERS = {"Content-Type": "text/html; charset=UTF-8"}

# <---------------------------------Result------------------------------------>

# Result (Single event page)
# Example: https://iwf.sport/results/results-by-events/?event_id=529
RESULT_URL = "event_id="

# <---------------------------------Events------------------------------------>

# Event
# Example: https://iwf.sport/results/results-by-events/
EVENTS_URL = "/results/results-by-events"
OLD_BW_EVENTS_URL = "/results/results-by-events/results-by-events-old-bw"

# Event Year
# Example: https://iwf.sport/results/results-by-events/?event_year=2022
YEAR_URL = "?event_year="

# Event Filter
# Example: https://iwf.sport/results/results-by-events/?event_type=all&event_age=all&event_nation=all
EVENT_TYPE_URL = "event_type="
EVENT_AGE_URL = "event_age="
EVENT_NATION_URL = "event_nation="


# <---------------------------------Athletes---------------------------------->

# Athlete
ATHLETE_URL = "/weightlifting_/athletes-bios/"

# Athlete Search
# Example: https://iwf.sport/weightlifting_/athletes-bios/?athlete_name=test&athlete_gender=all&athlete_nation=all
ATHLETE_NAME = "athlete_name="

# Athlete Filter
ATHLETE_GENDER = "athlete_gender="
ATHLETE_NATION = "athlete_nation="


# <---------------------------------Athlete-Bio------------------------------->

# Athlete Bio (Single athlete page)
# Example: https://iwf.sport/weightlifting_/athletes-bios/?athlete=ilyin-ilya-1988-05-24&id=7895
# Must have both
ATHLETE_BIO_URL = "athlete="
ATHLETE_ID_URL = "id="


TIME_TO_WAIT = 3  # TODO: consider using Selenium WebDriver Wait API


def is_event(url):
    """
    Validates events page
    Example: https://iwf.sport/results/results-by-events/?event_type=all&event_age=all&event_nation=all
    """
    return True if EVENTS_URL or OLD_BW_EVENTS_URL in url else False


def is_result(url):
    """
    Validates event url.
    Example: https://iwf.sport/results/results-by-events/?event_id=529
    """
    return True if RESULT_URL in url else False


def is_athlete_bio(url):
    """
    Validate athlete url.
    Example: https://iwf.sport/weightlifting_/athletes-bios/?athlete=ilyin-ilya-1988-05-24&id=7895
    """
    return True if ATHLETE_BIO_URL in url else False


# def _load_new_bodyweight_events_page():
#     """
#     Loads the page for new bodyweight category
#     """
#     r = requests.get(BASE_URL + EVENTS_URL, headers=HEADERS)
#     html = r.text
#     return BeautifulSoup(html, "lxml")


# def _load_old_bodyweight_events_page():
#     """
#     Loads the page for new bodyweight category
#     """
#     r = requests.get(BASE_URL + OLD_BW_EVENTS_URL, headers=HEADERS)
#     html = r.text
#     return BeautifulSoup(html, "lxml")


def _scrape_select_years(page):
    """
    Scrapes data for new or old bodyweight page
    """
    select_option = page.findAll("select", {"name": "event_year"})[0]
    options = select_option.findAll("option")
    years = []
    for item in options:
        years.append(item.get_text())

    return years


def get_years():
    """
    Gets all years available.
    New bodyweight years not needed since old bodyweight <select> includes them
    """
    old_events_years = _scrape_select_years(_load_old_bodyweight_events_page())
    return old_events_years
